pkg/k8s/watchers/cilium_network_policy.go

```diff
@@ -271,7 +271,7 @@ func (k *K8sWatcher) onUpsert(
 		err = k.addCiliumNetworkPolicyV2(cs, translatedCNP, initialRecvTime, resourceID)
 	}
 	if err == nil {
-		cnpCache[key] = cnpCpy
+		cnpCache[key] = cnp
```

Comment:
I think we depend on the copy to be in the cache, right @pippolo84?

Reply:
Why?

Reply:
AFAIU this race is happening because in case of a policy without any CIDRGroup reference, the function `resolveCIDRGroupRef` returns the same pointer without actually copying the cnp (for efficiency reasons). IOW we might have `translatedCNP == cnpCpy` after `resolveCIDRGroupRef` is called.
Then, `translatedCNP` is a parameter in `k.addCiliumNetworkPolicyV2` where, in case of `option.Config.DisableCNPStatusUpdates == false`, it'll be captured by closure in the controller (and thus written by the goroutine later). At the same time, `cnpCpy` will be saved in the cache and later read to compare it with `DeepEqual()` to the upcoming event from the CNP resource watcher.

I think we were storing `cnpCpy` in the cache because that can be written later. It is passed as a parameter (`oldCNP`) in `k.updateCiliumNetworkPolicyV2`, where the rules sanitization calls the `(*IngressCommonRule) SetAggregatedSelectors()` method, that might rewrite the rules.
I mentioned this in the last part of this [comment](https://github.com/cilium/cilium/pull/24638#discussion_r1155786399) in the original PR.
So, I don't think we can save `cnp` in the cache, because that is coming from the CNP resource underlying Informer and AFAIK cannot be written.

What about making a deep copy of the `cnp` in `(c *CNPStatusUpdateContext) updateViaAPIServer` instead?

Reply:
@aanm It's a bit subtle because of the CIDRGroup feature but what Fabio describes sounds about right. I think this cache thingy is quite fragile so we should improve it but not necessarily in this PR. I think a code comment similar to what Fabio posted above would be good in the meantime to warn developers about depending on the copy.

Reply:
@christarazi @pippolo84 what if we add the following:
```golang
	oldCNP, ok := cnpCache[key]
	if ok {
		if oldCNP.DeepEqual(cnp) {
			return nil
		}
		oldCNPCpy := oldCNP.DeepCopy()
		oldCNP = oldCNPCpy
	}
```
between lines 236 and 241?

Reply:
This should work fine. The only downside is that we're gonna pay the overhead of an extra copy every time there is a cache hit, even when `option.Config.DisableCNPStatusUpdates == true`.

I think the following should work fine either with less overhead when CNP StatusUpdates are disabled:

```diff
@@ -349,7 +349,9 @@ func (k *K8sWatcher) addCiliumNetworkPolicyV2(ciliumNPClient clientset.Interface
                        controller.ControllerParams{
                                Group: syncCNPStatusControllerGroup,
                                DoFunc: func(ctx context.Context) error {
-                                       return updateContext.UpdateStatus(ctx, cnp, rev, policyImportErr)
+                                       // We need to deep-copy the CNP object because UpdateStatus modifies it
+                                       // to clear the LastAppliedConfiguration key from annotations map.
+                                       return updateContext.UpdateStatus(ctx, cnp.DeepCopy(), rev, policyImportErr)
                                },
                        },
                )
@@ -485,7 +487,9 @@ func (k *K8sWatcher) updateCiliumNetworkPolicyV2AnnotationsOnly(ciliumNPClient c
                controller.ControllerParams{
                        Group: syncCNPStatusControllerGroup,
                        DoFunc: func(ctx context.Context) error {
-                               return updateContext.UpdateStatus(ctx, cnp, meta.revision, meta.policyImportError)
+                               // We need to deep-copy the CNP object because UpdateStatus modifies it
+                               // to clear the LastAppliedConfiguration key from annotations map.
+                               return updateContext.UpdateStatus(ctx, cnp.DeepCopy(), meta.revision, meta.policyImportError)
                        },
                })
```